tests/test_json.py

```diff
@@ -0,0 +1,190 @@
+    json_writer.writer(new_file, schema, records)
+    new_file.seek(0)
+
+    new_records = list(json_reader.reader(new_file, schema))
+    return new_records
+
+
+def test_json():
+    schema = {
+        "type": "record",
+        "name": "Test",
+        "namespace": "test",
+        "fields": [{
+            "name": "null",
+            "type": "null",
+        }, {
+            "name": "boolean",
+            "type": "boolean",
+        }, {
+            "name": "string",
+            "type": "string",
+        }, {
+            "name": "bytes",
+            "type": "bytes",
+        }, {
+            "name": "int",
+            "type": "int",
+        }, {
+            "name": "long",
+            "type": "long",
+        }, {
+            "name": "float",
+            "type": "float",
+        }, {
+            "name": "double",
+            "type": "double",
+        }, {
+            "name": "fixed",
+            "type": {
+                "type": "fixed",
+                "name": "fixed_field",
+                "size": 5,
+            },
+        }, {
+            "name": "union",
+            "type": [
+                'null',
+                'int',
+                {
+                    "type": "record",
+                    "name": "union_record",
+                    "fields": [{
+                        "name": "union_record_field",
+                        "type": "string",
+                    }],
+                },
+            ]
+        }, {
+            "name": "enum",
+            "type": {
+                "type": "enum",
+                "symbols": ["FOO", "BAR"],
+            },
+        }, {
+            "name": "array",
+            "type": {
+                "type": "array",
+                "items": "string",
+            },
+        }, {
+            "name": "map",
+            "type": {
+                "type": "map",
+                "values": "int",
+            },
+        }, {
+            "name": "record",
+            "type": {
+                "type": "record",
+                "name": "subrecord",
+                "fields": [{
+                    "name": "sub_int",
+                    "type": "int",
+                }],
+            },
+        }]
+    }
+
+    records = [{
+        'null': None,
+        'boolean': True,
+        'string': 'foo',
+        'bytes': b'\xe2\x99\xa5',
+        'int': 1,
+        'long': 1 << 33,
+        'float': 2.2,
+        'double': 3.3,
+        'fixed': b'\x61\x61\x61\x61\x61',
+        'union': None,
+        'enum': 'BAR',
+        'array': ['a', 'b'],
+        'map': {
+            'c': 1,
+            'd': 2
+        },
+        'record': {
+            'sub_int': 123,
+        }
+    }, {
+        'null': None,
+        'boolean': True,
+        'string': 'foo',
+        'bytes': b'\xe2\x99\xa5',
+        'int': 1,
+        'long': 1 << 33,
+        'float': 2.2,
+        'double': 3.3,
+        'fixed': b'\x61\x61\x61\x61\x61',
+        'union': 321,
```

Comment:
This is wrong. The correct datum would be:
```
'union': {  'int': 321}
```

Reply:
So, the main reason why I asked modification in fastavro itself is - even if I write my wrapper, as I did, to handle the json encoding, when you transform from json encoding to non json encoding you are losing information and that's the part we don't want. 

Unfortunately, I see little value in keeping sending this to `json_writer.writer`.  What we really need is probably to change the c-level library to make it accept json encoding and do the opposite, make the old methods to convert from non-json enconding to json encoding, by making guesses and inferring things.

Reply:
Actually, thinking on the float/double/fixed use case above, it would be really complicated to send this to C. 
Maybe what should be sent to the C library is neither json spec nor current implementation, is a specific protocol format that both json encoding and normal encoding should be converted to/from.